qbft/state.go

```diff
@@ -88,7 +88,7 @@ type State struct {
 	Height                          Height
 	LastPreparedRound               Round
 	LastPreparedValue               []byte
-	ProposalAcceptedForCurrentRound *types.SignedSSVMessage
+	ProposalAcceptedForCurrentRound *ProcessingMessage
```

Comment:
must we have a state change? This just changes all tests...
If we must it is fine

Reply:
I can revert to `types.SignedSSVMessage` but this will add again the decoding (only for this variable). What do you think? I have no opinion here

Reply:
where do you do the decoding?
I can't find it

Reply:
I mean that if we store `ProposalAcceptedForCurrentRound` as a `types.SignedSSVMessage` then once we use it (e.g. for UponPrepare and UponCommit) we will need to decode its content again

Reply:
Here is the old `UponCommit`:
```go
// UponCommit returns true if a quorum of commit messages was received.
// Assumes commit message is valid!
func (i *Instance) UponCommit(signedCommit *types.SignedSSVMessage, commitMsgContainer *MsgContainer) (bool, []byte, *types.SignedSSVMessage, error) {
	// Decode qbft message
	msg, err := DecodeMessage(signedCommit.SSVMessage.Data)
	if err != nil {
		return false, nil, nil, err
	}

	addMsg, err := commitMsgContainer.AddFirstMsgForSignerAndRound(signedCommit)
	if err != nil {
		return false, nil, nil, errors.Wrap(err, "could not add commit msg to container")
	}
	if !addMsg {
		return false, nil, nil, nil // UponCommit was already called
	}

	// calculate commit quorum and act upon it
	quorum, commitMsgs, err := commitQuorumForRoundRoot(i.State, commitMsgContainer, msg.Root, msg.Round)
	if err != nil {
		return false, nil, nil, errors.Wrap(err, "could not calculate commit quorum")
	}
	if quorum {
		fullData := i.State.ProposalAcceptedForCurrentRound.FullData /* must have value there, checked on validateCommit */

		agg, err := aggregateCommitMsgs(commitMsgs, fullData)
		if err != nil {
			return false, nil, nil, errors.Wrap(err, "could not aggregate commit msgs")
		}
		return true, fullData, agg, nil
	}
	return false, nil, nil, nil
}
```

I don't see us decoding

Reply:
In the current alan branch:
```go
func (i *Instance) uponPrepare(signedPrepare *types.SignedSSVMessage, prepareMsgContainer *MsgContainer) error {
	// ...

	if !HasQuorum(i.State.CommitteeMember, prepareMsgContainer.MessagesForRound(i.State.Round)) {
		return nil // no quorum yet
	}
        // > Here
	proposalMsgAccepted, err := DecodeMessage(i.State.ProposalAcceptedForCurrentRound.SSVMessage.Data)
	if err != nil {
		return err
	}

	proposedRoot := proposalMsgAccepted.Root
	// ...
}

func (i *Instance) BaseMsgValidation(signedMsg *types.SignedSSVMessage) error {
        // ...
        case PrepareMsgType:
		proposedSignedMsg := i.State.ProposalAcceptedForCurrentRound
		if proposedSignedMsg == nil {
			return errors.New("did not receive proposal for this round")
		}

                // > Here
		proposedMsg, err := DecodeMessage(proposedSignedMsg.SSVMessage.Data)
		if err != nil {
			return errors.Wrap(err, "proposal saved for this round is invalid")
		}

		return validSignedPrepareForHeightRoundAndRootIgnoreSignature(
			signedMsg,
			i.State.Height,
			i.State.Round,
			proposedMsg.Root,
			i.State.CommitteeMember.Committee,
		)
	case CommitMsgType:
		proposedMsg := i.State.ProposalAcceptedForCurrentRound
		if proposedMsg == nil {
			return errors.New("did not receive proposal for this round")
		}
		return validateCommit(
			signedMsg,
			i.State.Height,
			i.State.Round,
			i.State.ProposalAcceptedForCurrentRound,
			i.State.CommitteeMember.Committee,
		)
	case RoundChangeMsgType:
	// ...
}
```